plugins/kinetic-dp/fu-kinetic-dp-firmware.c

```diff
@@ -0,0 +1,259 @@
+	firmware->cmdb_block_size = _get_valid_payload_size (&fw_bin_buf[SPI_CMDB_BLOCK_START], CMDB_BLOCK_SIZE);
+
+	return TRUE;
+}
+
+guint32
+fu_kinetic_dp_firmware_get_esm_payload_size (FuKineticDpFirmware *self)
+{
+	g_return_val_if_fail (FU_KINETIC_DP_FIRMWARE (self), 0);
+	return self->esm_payload_size;
+}
+
+guint32
+fu_kinetic_dp_firmware_get_arm_app_code_size (FuKineticDpFirmware *self)
+{
+	g_return_val_if_fail (FU_KINETIC_DP_FIRMWARE (self), 0);
+	return self->arm_app_code_size;
+}
+
+guint32
+fu_kinetic_dp_firmware_get_app_init_data_size (FuKineticDpFirmware *self)
+{
+	g_return_val_if_fail (FU_KINETIC_DP_FIRMWARE (self), 0);
+	return self->app_init_data_size;
+}
+
+guint32
+fu_kinetic_dp_firmware_get_cmdb_block_size (FuKineticDpFirmware *self)
+{
+	g_return_val_if_fail (FU_KINETIC_DP_FIRMWARE (self), 0);
+	return self->cmdb_block_size;
+}
+
+gboolean
+fu_kinetic_dp_firmware_get_is_fw_esm_xip_enabled (FuKineticDpFirmware *self)
+{
+	g_return_val_if_fail (FU_KINETIC_DP_FIRMWARE (self), FALSE);
+	return self->is_fw_esm_xip_enabled;
+}
+
+static gboolean
+fu_kinetic_dp_firmware_parse (FuFirmware *firmware,
+			      GBytes *fw,
+			      guint64 addr_start,
+			      guint64 addr_end,
+			      FwupdInstallFlags flags,
+			      GError **error)
+{
+	FuKineticDpFirmware *fw_self = FU_KINETIC_DP_FIRMWARE (firmware);
+	const guint8 *buf;
+	gsize bufsz;
+	guint32 isp_drv_payload_size = 0, app_fw_payload_size = 0;
+	g_autoptr (GBytes) isp_drv_payload = NULL;
+	g_autoptr (GBytes) app_fw_payload = NULL;
+	g_autoptr  (FuFirmware) isp_drv_img = NULL;
+	g_autoptr (FuFirmware) app_fw_img = NULL;
+	KtChipId chip_id = KT_CHIP_NONE;	// <TODO> store in class private data
+	guint16 fw_bin_flag = 0;		// <TODO> store in class private data
+
+	/* Parse firmware according to Kinetic's FW image format
+	 * FW binary = 4 bytes header(Little-Endian) + ISP driver + app FW
+	 * 4 bytes: size of ISP driver
+	 */
+	buf = g_bytes_get_data (fw, &bufsz);
+	if (!fu_common_read_uint32_safe(buf, bufsz, 0,
+					&isp_drv_payload_size, G_LITTLE_ENDIAN, error)) {
+		return FALSE;
+	}
+	g_debug ("ISP driver payload size: %u bytes", isp_drv_payload_size);
+
+	app_fw_payload_size = g_bytes_get_size (fw) - HEADER_LEN_ISP_DRV_SIZE - isp_drv_payload_size;
+	g_debug ("App FW payload size: %u bytes", app_fw_payload_size);
+
+	/* Add ISP driver as a new image into firmware */
+	isp_drv_payload = g_bytes_new_from_bytes (fw, HEADER_LEN_ISP_DRV_SIZE, isp_drv_payload_size);
+	isp_drv_img = fu_firmware_new_from_bytes (isp_drv_payload);
+	fu_firmware_set_idx (isp_drv_img, FU_KT_FW_IMG_IDX_ISP_DRV);
+
+	fu_firmware_add_image (firmware, isp_drv_img);
+
+	/* Add App FW as a new image into firmware */
+	app_fw_payload = g_bytes_new_from_bytes (fw, HEADER_LEN_ISP_DRV_SIZE + isp_drv_payload_size, app_fw_payload_size);
+	buf = g_bytes_get_data (app_fw_payload, &bufsz);
+	if (!kt_dp_get_chip_id_from_fw_buf (buf, bufsz, &chip_id, &fw_bin_flag)) {
```

Comment:
you need to send error in this function and set it on the FALSE scenario

Reply:
> some general feedback:
> 
> * There are a lot of `g_message` commands, none of these should stay.  They should either be removed (now that you've finished bringup) or they should downgrade to `g_debug`.  Up to you on which one, but please don't leave them.
> * There are various TODOs left, do you want any of these to block merging?  Or they're just nice to do?
> * Please update comments style to `/* */` instead of `//`
> * All of your instances of `g_prefix_error` that end in `!` should correct to end with `: ` so that they prefix properly

I've fixed most of the feedback in the new commit.

And I don't want the TODO list to block merging.
I think I would have some issues to implement after I can do online test.
As you said, some of the TODOs are just nice to do.
I'll complete them if the implementation details are determined.

Thanks.